File: 백준/Gold/2504. 괄호의 값/괄호의 값.py
import sys
input = sys.stdin.readline

munja = input().strip()

stack = []

for i in range(len(munja)):
    if munja[i] == "(" or munja[i] == "[":
        stack.append(munja[i])
    elif munja[i] == ")":
        if not stack:
            print(0)
            sys.exit()
        tmp = stack.pop()
        if tmp == "(":
            stack.append(2)
        elif tmp == "[":
            print(0)
            sys.exit()
        else:
            res = int(tmp)
            while stack and stack[-1] != "(":
                if stack[-1] == "[":
                    print(0)
                    sys.exit()
                res += stack.pop()
            if stack:
                stack.pop()
                stack.append(res * 2)
            else:
                print(0)
                sys.exit()
    elif munja[i] == "]":
        if not stack:
            print(0)
            sys.exit()
        tmp = stack.pop()
        if tmp == "[":
            stack.append(3)
        elif tmp == "(":
            print(0)
            sys.exit()
        else:
            res = int(tmp)
            while stack and stack[-1] != "[":
                if stack[-1] == "(":
                    print(0)
                    sys.exit()      
                res += stack.pop()
            if stack:
                stack.pop()
                stack.append(res * 3)
            else:
                print(0)
                sys.exit()
for i in stack:
    if i == "(" or i == "[":
        print(0)
        sys.exit()

print(sum(stack))

File: 백준/Gold/2504. 괄호의 값/괄호의 값.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String s = br.readLine();

        Stack<Character> stack = new Stack<>();
        int ans = 0;
        int tmp = 1;

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);

            if (c == '(') {
                stack.push(c);
                tmp *= 2;
            } else if (c == '[') {
                stack.push(c);
                tmp *= 3;
            } else if (c == ')') {
                if (stack.isEmpty() || stack.peek() != '(') {
                    System.out.println(0);
                    return;
                }
                // "()"면 현재 tmp를 더함
                if (i > 0 && s.charAt(i - 1) == '(') {
                    ans += tmp;
                }
                stack.pop();
                tmp /= 2;
            } else if (c == ']') {
                if (stack.isEmpty() || stack.peek() != '[') {
                    System.out.println(0);
                    return;
                }
                
                if (i > 0 && s.charAt(i - 1) == '[') {
                    ans += tmp;
                }
                stack.pop();
                tmp /= 3;
            }
        }

        if (!stack.isEmpty()) {
            System.out.println(0);
        } else {
            System.out.println(ans);
        }
    }
}